### PR TITLE
feat: peekable and parallel dataset

### LINKED ISSUE #202: add parallel pipeline for order by sort.
**What would you like to be added**:

**Why is this needed**:

### PR DESCRIPTION
**Pull request by jjeffcaii**
**What this PR does**: implement  parallel and peekable dataset.

**Which issue(s) this PR fixes**:

Fixes #202 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by dongzl — APPROVED** [review]

File: pkg/dataset/parallel.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+	if next, err, at.next, at.err = at.next, at.err, nil, nil; next != nil || err != nil {
+		return next, err
+	}
+
+	return at.Dataset.Next()
+}
+
+func (at *peekableDataset) Peek() (proto.Row, error) {
+	at.mu.Lock()
```
Comment (AlexStocks) [inline pkg/dataset/parallel.go:65]: 按照传统的C习惯，把 变量定义放在代码最开始部分

File: pkg/dataset/parallel.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+type peekableDataset struct {
+	proto.Dataset
+	mu   sync.Mutex
+	next proto.Row
+	err  error
+}
+
+func (at *peekableDataset) Next() (proto.Row, error) {
+	at.mu.Lock()
```
Comment (AlexStocks) [inline pkg/dataset/parallel.go:49]: 按照传统的C习惯，把 变量定义放在代码最开始部分

File: pkg/dataset/parallel.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+		pa.mu.RUnlock()
+		return stream, nil
+	}
+
+	pa.mu.RUnlock()
+
+	pa.mu.Lock()
+	defer pa.mu.Unlock()
+
+	if stream := pa.streams[i]; stream != nil {
```
Comment (georgehao) [inline pkg/dataset/parallel.go:102]: retry get the stream, seems no need

Comment (georgehao) [inline pkg/dataset/parallel.go:102]: retry get stream is no need

**Review by georgehao — APPROVED** [review]

**Review by baerwang — APPROVED** [review]

**Review by Lvnszn — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**